gt/main.c

```diff
@@ -51,6 +63,66 @@ gt_setup_rss(struct gt_config *gt_conf)
 }
 
 static int
+extract_packet_fields(struct rte_mbuf *m, struct gt_packet_fields *fields)
+{
+	struct ipv4_hdr *ip4_hdr = rte_pktmbuf_mtod(m, struct ipv4_hdr *);
+	uint8_t inner_ip_ver = (ip4_hdr->version_ihl & 0xF0) >> 4;
+	fields->l3_hdr = ip4_hdr;
+
+	if (inner_ip_ver == 4) {
+		fields->inner_ip_ver = ETHER_TYPE_IPv4;
+		fields->l4_proto = ip4_hdr->next_proto_id;
+		fields->l4_hdr = &ip4_hdr[1];
```

Comment:
This cast should verify that the packet has space for at least 4 bytes for the l4 header. The Lua policy will be responsible for checking the necessary space for each type. You should document this fact in the example policy.